File: src/main/core/playtime.js
'use strict';

const { JsonStore } = require('./store');

/**
 * Учёт игрового времени — как в Steam, GOG Galaxy и Playnite (1.10).
 *
 * Считается то, что запущено кнопкой «Играть» в ModHub: от старта процесса
 * игры до его завершения. Часть игр при запуске мимо Steam перезапускает
 * себя через Steam — такой процесс живёт пару секунд, и эти секунды за
 * сеанс не считаются: лучше недосчитать, чем записать нули и мусор.
 */

const MIN_SESSION_MS = 15_000;
/** Сеанс дольше суток — скорее всего спящий компьютер или сбитые часы. */
const MAX_SESSION_MS = 24 * 60 * 60 * 1000;

class PlayTime {
  /** @param {{file: string, now?: () => number}} options */
  constructor({ file, now = () => Date.now() }) {
    this.store = new JsonStore(file, { games: {} });
    this.now = now;
    this.running = new Map(); // gameId -> startedAt
  }

  /** Игра запущена. */
  start(gameId) {
    const at = this.now();
    this.running.set(gameId, at);
    const game = this._game(gameId);
    game.lastPlayed = new Date(at).toISOString();
    this.store.save();
    return at;
  }

  /**
   * Процесс игры завершился.
   * @returns {{counted: boolean, ms: number}}
   */
  stop(gameId) {
    const startedAt = this.running.get(gameId);
    if (startedAt === undefined) return { counted: false, ms: 0 };
    this.running.delete(gameId);
    const ms = this.now() - startedAt;
    if (ms < MIN_SESSION_MS || ms > MAX_SESSION_MS) return { counted: false, ms };
    const game = this._game(gameId);
    game.totalMs += ms;
    game.sessions += 1;
    game.lastSessionMs = ms;
    game.longestMs = Math.max(game.longestMs ?? 0, ms);
    this.store.save();
    return { counted: true, ms };
  }

  isRunning(gameId) {
    return this.running.has(gameId);
  }

  _game(gameId) {
    if (!this.store.data.games[gameId]) {
      this.store.data.games[gameId] = { totalMs: 0, sessions: 0, lastPlayed: null, lastSessionMs: 0, longestMs: 0 };
    }
    return this.store.data.games[gameId];
  }

  /** Всё по всем играм: { [gameId]: { totalMs, sessions, lastPlayed, running } }. */
  all() {
    const out = {};
    for (const [id, g] of Object.entries(this.store.data.games)) {
      out[id] = { ...g, running: this.running.has(id) };
    }
    for (const id of this.running.keys()) {
      if (!out[id]) out[id] = { totalMs: 0, sessions: 0, lastPlayed: null, running: true };
    }
    return out;
  }

  reset(gameId) {
    if (gameId) delete this.store.data.games[gameId];
    else this.store.data.games = {};
    this.store.save();
    return true;
  }
}

module.exports = { PlayTime, MIN_SESSION_MS };

File: desktop/ModLaunch/Features/PlayTime.cs
using System.Text.Json.Nodes;
using ModLaunch.Core;

namespace ModLaunch.Features;

public sealed record Played(long TotalMs, int Sessions, DateTime? LastPlayed, bool Running);

/// <summary>Игровое время: playtime.json, как в 3.x. Сеансы короче 15 секунд не считаются.</summary>
public static class PlayTime
{
    static readonly JsonFile File = new(Path.Combine(Paths.DataDir, "playtime.json"), () => new JsonObject { ["games"] = new JsonObject() });
    static readonly Dictionary<string, DateTime> RunningSince = [];

    public static bool Track => Settings.Data.Bool("trackPlaytime", true);

    static JsonObject Game(string id)
    {
        var games = File.Data.Obj("games");
        if (games[id] is JsonObject g) return g;
        g = new JsonObject { ["totalMs"] = 0, ["sessions"] = 0, ["lastPlayed"] = null, ["lastSessionMs"] = 0, ["longestMs"] = 0 };
        games[id] = g;
        return g;
    }

    public static bool IsRunning(string id) => RunningSince.ContainsKey(id);

    public static void Start(string id)
    {
        RunningSince[id] = DateTime.UtcNow;
        Game(id)["lastPlayed"] = DateTime.UtcNow.ToString("o");
        File.Save();
    }

    public static (bool Counted, long Ms) Stop(string id)
    {
        if (!RunningSince.Remove(id, out var since)) return (false, 0);
        var ms = (long)(DateTime.UtcNow - since).TotalMilliseconds;
        if (ms < 15_000 || ms > 86_400_000) return (false, ms);
        var g = Game(id);
        g["totalMs"] = g.Long("totalMs") + ms;
        g["sessions"] = g.Long("sessions") + 1;
        g["lastSessionMs"] = ms;
        g["longestMs"] = Math.Max(g.Long("longestMs"), ms);
        File.Save();
        return (true, ms);
    }

    public static Played Get(string id)
    {
        var g = File.Data.Obj("games")[id];
        return new Played(g.Long("totalMs"), (int)g.Long("sessions"),
            DateTime.TryParse(g.Str("lastPlayed"), out var d) ? d.ToUniversalTime() : null, IsRunning(id));
    }

    public static void Reset(string? id)
    {
        if (id is null) File.Data["games"] = new JsonObject(); else File.Data.Obj("games").Remove(id);
        File.Save();
    }

    public static string Format(long ms)
    {
        var minutes = ms / 60000;
        if (minutes < 1) return I18n.T("time.lessMinute");
        if (minutes < 60) return I18n.T("time.m", ("m", minutes));
        var h = minutes / 60;
        var m = minutes % 60;
        return m == 0 ? I18n.T("time.h", ("h", h)) : I18n.T("time.hm", ("h", h), ("m", m));
    }
}
